include/mount.h

```diff
@@ -0,0 +1,100 @@
+#ifndef _SYS_MOUNT_H_
+#define _SYS_MOUNT_H_
+
+#include <queue.h>
+#include <mutex.h>
+
+/* Maximum length of a filesystem type name */
+#define VFSCONF_NAME_MAX 32
+/* Maximum length of a path to lookup */
+#define VFS_PATH_MAX 256
+
+/* Forward declarations */
+typedef struct vnode vnode_t;
+typedef struct mount mount_t;
+typedef struct vnode vnode_t;
+typedef struct vfsconf vfsconf_t;
+typedef struct statfs statfs_t;
+
+/* VFS operations */
+typedef int vfs_mount_t(mount_t *m);
+typedef int vfs_root_t(mount_t *m, vnode_t **vp);
+typedef int vfs_statfs_t(mount_t *m, statfs_t *sb);
+typedef int vfs_vget_t(mount_t *m, ino_t ino, vnode_t **vp);
+typedef int vfs_init_t(vfsconf_t *vfc);
+
+typedef struct vfsops {
+  vfs_mount_t *vfs_mount;
+  vfs_root_t *vfs_root;
+  vfs_statfs_t *vfs_statfs;
+  vfs_vget_t *vfs_vget;
+  vfs_init_t *vfs_init;
+} vfsops_t;
+
+/* Description of a filesystem type. There is one instance of this struct per
+ * each filesystem type in the kernel (e.g. tmpfs, devfs). */
+typedef struct vfsconf {
+  char vfc_name[VFSCONF_NAME_MAX]; /* Filesystem type name */
+  vfsops_t *vfc_vfsops;            /* Filesystem operations */
+  int vfc_refcnt; /* Number of mounted filesystems of this type */
```

Comment:
Note that this is *not* a reference count, and naming it `refcnt` may be confusing. This field is used to count the number of filesystem instances of this type - the number of actual references may be higher. In particular, when a file system type is registered but not mounted anywhere, this counter stays at 0 - even though VFS does have a reference to this `vfsconf`. Also since there is no need to ever free `vfsconfs`, a reference counter wouldn't be useful.